Automated release workflow

trask opened the pull request:
Hey all!

We've spent a lot of time lately improving our release workflow in the OTel Java repos, and I'd like to see if our toil and pain can be of any benefit to other repos (starting with python of course!).

I also think there's some benefit of shared cross-SIG knowledge going forward if we follow somewhat similar workflow practices.

I've tried to document things here: https://github.com/trask/repository-template/blob/main/template-docs/release-automation.md

Some highlights in this PR:
* Uses `-dev` version suffix on `main` (@lzchen mentioned you were thinking about doing this anyways)
* The release workflow autogenerates the release notes from the change log and publishes the github release
* Uses the bot workaround discussed in https://github.com/open-telemetry/community/issues/809#issuecomment-1015121722 (maybe we should consider creating a shared otel bot that all repos could use?)
* The branch names are based on the minor release (e.g. `release/v1.12.x-0.31bx`), and patch releases are made out of those same branches. Tags are still applied to each individual release.

New complications in this repo that we hadn't faced in the Java repos:
* Separate versions for stable/unstable components (this wasn't a big deal, just a bit more workflow code to handle/increment them both).
* I tried adding the `Skip Changelog` label to the generated pull requests that target `main`, but that requires giving the bot triage rights to the repo, so instead suppressed the changelog workflow when `github.actor` is the bot.

I'd love to stop by your SIG meeting, but it's always at the same time as the Java SIG 😢. Maybe next week I can sneak away and jump to your meeting for a bit.

Please throw your questions this way though, I'd love to better understand where we can align more, and where things will naturally diverge based on SIG needs/preferences.

trask commented on CHANGELOG.md:
```diff
@@ -24,7 +23,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2583](https://github.com/open-telemetry/opentelemetry-python/pull/2583))
 
-## [1.11.0-0.30b0](https://github.com/open-telemetry/opentelemetry-python/releases/tag/v1.11.0-0.30b0) - 2022-04-18
+## Version 1.11.0/0.30b0 (2022-04-18)
```
these links didn't seem to work anyways, so I took the liberty to simplify/align with what we are doing in the Java repos, but I can bring them back and update the release automation regexes accordingly if you'd like

trask commented:
I should have mentioned, I tested all of these workflows on my python repo fork, though there may still be a hiccup or two on the first try here. I'd be happy to coordinate and be online with you during the first release using these new workflows if you are interested.

srikanthccv commented:
Sorry I haven't gotten a time to review this in detail but there are many things I like about this. Team has been focused on getting the rc for metrics work for the past few weeks. So it might take sometime before we review and get it merged.

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+* Merge a pull request to `main` updating the `CHANGELOG.md`.
+  * The heading for the release should include the release version but not the release date, e.g.
+    `## Version 1.9.0 (unreleased)`.
+  * Use `.github/scripts/draft-change-log-entries.sh` as a starting point for writing the change
```
I can't seem to find this file?

trask replied:
I removed this bullet since you are building your change log as you go in this repo

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+    `## Version 1.9.0 (unreleased)`.
+  * Use `.github/scripts/draft-change-log-entries.sh` as a starting point for writing the change
+    log.
+* Run the [Prepare release branch workflow](https://github.com/trask/repository-template/actions/workflows/prepare-release-branch.yml).
```
Should this link to this repositories workflow?

trask replied:
fixed

lzchen commented on .github/workflows/prepare-release-branch.yml:
```diff
@@ -0,0 +1,100 @@
+          echo "RELEASE_BRANCH_NAME=$release_branch_name" >> $GITHUB_ENV
+
+      - name: Update version
+        run: .github/scripts/update-version.sh $STABLE_VERSION $PRERELEASE_VERSION
```
Is there a reason why we can't run `./scripts/eachdist.py update_versions --versions stable,prerelease` directly from here instead of having to call a bash script?

trask replied:
You could definitely copy the 3-line shell script directly into the yaml. I primarily separated it into a script so that the `yml` files would be more similar across repos that adopt this (since the mechanics of updating versions themselves will vary a lot between repos, but the other parts of the yml do not vary much).

lzchen replied:
I was more referring to calling the python file directly in the `yml` file. The less scripting we can use the better (we want as many things to be as "pythonic" as possible. But not a blocking comment.

lzchen commented on .github/workflows/prepare-release-branch.yml:
```diff
@@ -0,0 +1,100 @@
+        run: .github/scripts/update-version.sh $STABLE_VERSION $PRERELEASE_VERSION
+
+      - name: Set git user
+        run: .github/scripts/set-git-user.sh
```
Curious, what is the purpose for setting the user? Is this for permissions to push to origin branches?

trask replied:
This is needed so that the PRs will pass the EasyCLA check, see https://github.com/open-telemetry/community/issues/809#issuecomment-1015121722

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+  * Press the "Run workflow" button, and leave the default branch `main` selected.
+* Review and merge the two pull requests that it creates
+  (one is targeted to the release branch and one is targeted to `main`).
+
```
```suggestion
   * These pull requests will update the release branch to the latest release versions and will bump the main branch to the next dev version.
```

trask replied:
(manually accepted)

lzchen commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,99 @@
+          # while release notes render them as line breaks
+          sed -n "0,/^## Version $STABLE_VERSION\/$PRERELEASE_VERSION/d;/^## Version /q;p" CHANGELOG.md \
+            | perl -0pe 's/(?<!\n)\n *(?!\n)(?![-*] )(?![1-9]+\. )/ /g' \
+            >> release-notes.txt
```
Will there always be a `release-notes.txt` in the repo? Should we do a cleanup after the release?

lzchen commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,99 @@
+        run: |
+          # conditional block not indented because of the heredoc
+          if [[ ! -z $PRIOR_COMBINED_VERSION_WHEN_PATCH ]]; then
+          cat > release-notes.txt << EOF
```
Will there always be a `release-notes.txt` in the repo?

trask replied:
this file is never committed, maybe this script would be clearer if it used `/tmp/release-notes.txt`?

lzchen replied:
Yeah sgtm.

trask replied:
Updated

lzchen commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,99 @@
+          # while release notes render them as line breaks
+          sed -n "0,/^## Version $STABLE_VERSION\/$PRERELEASE_VERSION/d;/^## Version /q;p" CHANGELOG.md \
+            | perl -0pe 's/(?<!\n)\n *(?!\n)(?![-*] )(?![1-9]+\. )/ /g' \
+            >> release-notes.txt
```
Are these release notes generated BEFORE the release date is updated in the CHANGELOG?

trask replied:
before

lzchen commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,99 @@
+                            --discussion-category announcements \
+                            v$STABLE_VERSION
+
+      - name: Update the change log with the release date
```
Curious as to why the change log is updated with the release date AFTER the release has already been created? Don't we want to have the published package to have the release date reflected in the CHANGELOG?

trask replied:
The release date could be populated in the "prepare release branch" PR. The only downside of that is the actual release date might be later.

Maybe do both? Populate it in the "prepare release branch" PR, and in the post-release workflow, update the date if it happened to be later? Or maybe even fail to release if the release date doesn't match today's date, and have it push a PR to update the date at that point (before the release, so that the release date is always correct +/- small error in the release)?

lzchen replied:
> Populate it in the "prepare release branch" PR, and in the post-release workflow, update the date if it happened to be later?

I think this would be good. As for preventing the release, I think small +/- error is not a big deal.

trask replied:
ya, I like this, I updated the template-repository and this PR to update the date in the "prepare release branch" and "prepare patch release" workflows as part of the existing version bumping PRs

lzchen commented on .github/workflows/merge-change-log-to-main.yml:
```diff
@@ -0,0 +1,36 @@
+          body="Merge change log updates from \`$GITHUB_REF_NAME\`."
+          branch="merge-change-log-updates-from-${GITHUB_REF_NAME//\//-}"
+
+          git format-patch --stdout HEAD..origin/$GITHUB_REF_NAME CHANGELOG.md | git apply --3way
```
What does this line do exactly?

trask replied:
it does a 3-way merge of a single file (I added a comment here)

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
-
-- Create the GH release from the main branch, using a new tag for this micro version, e.g. `v0.7.0`
-- Copy the changelogs from all packages that changed into the release notes (and reformat to remove hard line wraps)
-
```
Perhaps we can add a TODO to automate the updating `stable` tag below as well?

trask replied:
The `stable` tag is automatically bumped as part of the workflow already (unless I misunderstood your question)

lzchen replied:
Could you link to the line in which this happens? I can't seem to find it.

trask replied:
ya, the `prepare-release-branch.yml` and `prepare-patch-release.yml` both call the `update-versions.sh` script, e.g. https://github.com/open-telemetry/opentelemetry-python/pull/2649/files#diff-69e7d09297eb3770a53d0e2385c3593d60acc3c787222c3809f6a540c01d498bR35

lzchen replied:
This would be referring to updating the latest [stable](https://github.com/open-telemetry/opentelemetry-python/releases/tag/stable) tag to the release commit right? We need this for readthedocs to render the latest document changes.

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+    `## Version 1.9.0 (unreleased)`.
+  * Use `.github/scripts/draft-change-log-entries.sh` as a starting point for writing the change
+    log.
+* Run the [Prepare release branch workflow](https://github.com/trask/repository-template/actions/workflows/prepare-release-branch.yml).
```
Should there be a step before this that informs to update `eachdist.ini` to the correct versions?

trask replied:
The versions in `eachdist.ini` are automatically updated as part of the workflows

lzchen replied:
Could you refer to the line in which this happens?

trask replied:
ya, https://github.com/open-telemetry/opentelemetry-python/pull/2649/files#diff-629de75a59c87199cb26eafcb3067377c3b88c4cbb9be6b7cace2d0d68f41183R31

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
-
-
 ## Check PyPI
 
```
Are we able to remove the "Update main" section below?

trask replied:
removed

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+  request against `main`.
+
 
 Release Process:
```
Should probably move this ToC to the top and add the header links.

trask replied:
I merged these post-release steps into the "After the release" section

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
-
-## Open a Pull Request
-
-The PR should be opened from the `release/<version>` branch created as part of running `prepare_release.sh` in the steps above.
```
We can probably delete `prepare_release.sh` if we decide to go with automatic workflow.

trask replied:
removed

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+
+## Preparing a new patch release
+
+* Backport pull request(s) to the release branch.
```
Under which circumstances do we do a patch release? Is this the same as "hotfix" procedure down below?

trask replied:
Patch release is any release that bumps the last digit, e.g. `1.10.1` is a patch release on top of `1.10.0`. I believe this is synonymous with hotfix.

Although it looks like your hotfix procedure is currently to only publish the affected packages:

> With administrator priviledges for PyPi, manually publish the affected packages

In Java we do a full publishing of all packages since that is simpler for us and works with the same automation as a regular release.

trask replied:
I (optimistically) removed the "Hotfix procedure" section

lzchen replied:
Hmm, it could be debated on whether we want to do a release of all packages if the hotfix only pertains to a few. Maybe we can do a separated "hotfix" release process in which we can specify which packages to release?

trask replied:
> Hmm, it could be debated on whether we want to do a release of all packages if the hotfix only pertains to a few.

I personally think it's more effort to do a targeted release of only a few packages, plus by always releasing all packages at the same versions, it's typically easier (at least in the Java world) for users to target all of their dependencies to the same matching version.

But I'm happy to bring back the hotfix release process section, just let me know.

lzchen replied:
Not a huge opinion from my side. I'm happy to test out this way for now and leave out the hotfix specific section.

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+## Preparing a new patch release
+
+* Backport pull request(s) to the release branch.
+  * Run the [Backport workflow](https://github.com/trask/repository-template/actions/workflows/backport.yml).
```
Curious as to why this is called `backport.yml`?

trask replied:
It backports a PR from the main branch to a release branch, so that PR can be included in a patch release

lzchen commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+
+* Backport pull request(s) to the release branch.
+  * Run the [Backport workflow](https://github.com/trask/repository-template/actions/workflows/backport.yml).
+  * Press the "Run workflow" button, then select the release branch from the dropdown list,
```
We can run this workflow on multiple PRs if needed correct?

trask replied:
you can run it multiple times for multiple PRs (the branch name includes the PR #, so there's no conflict)

if it's common for you to backport multiple PRs, we could enhance the workflow to take a comma-separated list

lzchen replied:
Maybe a PR for the future.

trask commented on RELEASING.md:
```diff
@@ -1,48 +1,63 @@
+  * Use `.github/scripts/draft-change-log-entries.sh` as a starting point for writing the change
+    log.
+* Run the [Prepare release branch workflow](https://github.com/trask/repository-template/actions/workflows/prepare-release-branch.yml).
+  * Press the "Run workflow" button, and leave the default branch `main` selected.
```
you probably don't see it in that repo b/c you don't have permissions to run workflows

![image](https://user-images.githubusercontent.com/218610/167758166-0adb943f-3408-482a-9c8d-9393a0dff801.png)

lzchen replied:
Oh I can see it now thanks.

trask commented:
@lzchen thanks for all the comments! I have made several updates and I think replied to all of your comments, happy to make further updates

srikanthccv commented:
@trask will it be possible for you to attend tomorrow python SIG meeting and present it to all?

trask commented:
> @trask will it be possible for you to attend tomorrow python SIG meeting and present it to all?

yes, I'll plan on it, would it work to join at 9:30am Pacific Time?

srikanthccv commented:
That should work

lzchen commented:
@trask 

Will be approving after the `rc` case has been confirmed to work.

trask commented:
I updated the workflow to work with "pre-release" versions, e.g. "1.13.0rc", and tested making a "pre-release" in https://github.com/trask/repository-template/releases/tag/v1.1.0rc.

I'll re-test the full workflow in my `opentelemetry-python` fork one last time once everything is approved, but first could use some help resolving the build error below:

> ERROR: Cannot install opentelemetry-instrumentation-requests==0.31b0, opentelemetry-sdk==1.13.0.dev0 and opentelemetry-semantic-conventions 0.32b0.dev0 (from /home/runner/work/opentelemetry-python/opentelemetry-python/opentelemetry-semantic-conventions) because these package versions have conflicting dependencies.
> 
> The conflict is caused by:
>     The user requested opentelemetry-semantic-conventions 0.32b0.dev0 (from /home/runner/work/opentelemetry-python/opentelemetry-python/opentelemetry-semantic-conventions)
>     opentelemetry-sdk 1.13.0.dev0 depends on opentelemetry-semantic-conventions==0.32b0-dev
>     opentelemetry-instrumentation-requests 0.31b0 depends on opentelemetry-semantic-conventions==0.31b0
> 
> To fix this you could try to:
> 1. loosen the range of package versions you've specified
> 2. remove package versions to allow pip attempt to solve the dependency conflict

srikanthccv commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,38 @@
+      - name: Set git user
+        run: .github/scripts/set-git-user.sh
+
+      - name: Create pull request
```
This creates a pull request against release branches?

trask replied:
ya, you select the release branch when you run the workflow. I just added a check above this to verify that it's not being run against `main`.

srikanthccv commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,38 @@
+        run: |
+          commit=$(gh pr view $NUMBER --json mergeCommit --jq .mergeCommit.oid)
+          title=$(gh pr view $NUMBER --json title --jq .title)
+          url=$(gh pr view $NUMBER --json url --jq .url)
```
Did you plan to use this `url` somewhere?

trask replied:
oops, nope, removed

srikanthccv commented on .github/workflows/merge-change-log-to-main.yml:
```diff
@@ -0,0 +1,44 @@
+      - name: Set git user
+        run: .github/scripts/set-git-user.sh
+
+        # this will fail if there have been conflicting change log updates introduced in main
```
When does this happen? We make PR author to update the CHANGELOG entries.

trask replied:
good point, this is probably going to fail a lot and we should think of another option.

I guess now that the date is getting stamped onto the change log during prepare-release-branch, this step should only be needed for patch releases.

I'll change this to just be a manual step of merging the patch release change log back into main unless anyone has a better idea.

srikanthccv commented on .github/workflows/prepare-patch-release.yml:
```diff
@@ -0,0 +1,71 @@
+    steps:
+      - run: |
+          if [[ ! $GITHUB_REF_NAME =~ ^release/[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            echo this workflow should only be run against long-term release branches
```
Having difficulty in visualising what does workflow look like for patch release. Can you help me understand this? Let's say some bug was discovered after 1.11.0 release? What would be the new patch release process?

trask replied:
just checking that you've seen the steps in the RELEASING.md under "Preparing a new patch release"? https://github.com/open-telemetry/opentelemetry-python/pull/2649/files#diff-87d6a57a759f2386bf9caf5309b9aa0b9ff0ca70f5e38696e114a1ccdb0e785dR15

srikanthccv replied:
Ah, I missed that README section. It clarifies my doubt. I'll need to play around with these workflows to get a good grasp.

srikanthccv commented:
Our current workflow for patch/minor/major release is pretty simple. There are few additional steps introduced in this which also eliminates few manual steps we do today. Overall LGTM, I want to test on this repo or fork repo once.

ocelotl commented on .github/scripts/set-git-user.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash -e
+
+# TODO update with opentelemetry-python-bot or a shared otel bot
+git config user.name opentelemetry-java-bot
```
Just trying to understand, does [this](https://github.com/open-telemetry/community/issues/809#issuecomment-1016522419) mean it is necessary to create an `opentelemetry-python-bot` Github user?

trask replied:
there are three options:
* create an `opentelemetry-python-bot`
* create a shared `opentelemetry-bot`
* use the `opentelemetry-java-bot` for now (I can share a security token for this user privately)

codeboten replied:
fwiw, i would prefer a shared bot that all sigs can use, but it would be good to have this discussion w/ the greater community. this could be a good topic for the maintainers call

trask replied:
hey @codeboten! I noticed that the @opentelemetry-bot user has been used in the opentelemetry-collector-contrib repo: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aissue+author%3Aopentelemetry-bot+is%3Aclosed, do you happen to know who owns or has access to that? I would love to get that registered at https://github.com/open-telemetry/community/blob/main/assets.md, or we can come up with another bot name if that one is lost / unavailable for sharing.

trask commented:
oops! didn't mean to delete _that_ branch 😅

I made some nice improvements to the change log management based on conversations above.

I'm re-testing all the workflows on my fork right now and will post back when it's ready for another review

trask commented:
I tested the release workflow in my fork for the next release, which I understand will be 1.12.0rc2/0.32b0 (it could also be 0.31b1, but it makes the release workflow cleaner if you reserve b1 for patch releases).

I will test the patch release workflow tomorrow (e.g. I'll simulate 1.12.0 and then a 1.12.1 on the 1.12.x release branch).

btw, check out this section I added to `RELEASING.md` to clarify the situation with prereleases:

> ## Notes about "pre-releases"
>
> * Pre-release versions (e.g. `1.9.0rc1`) are supported, and will cause a "short-term" release branch to be created
  based on the full version name (e.g. `release/v1.9.0rc1` instead of a "long-term" release branch name like
  `release/v1.9.x`).
> * Patch releases are not supported on short-term release branches.
> * The version in `main` in this case will be bumped to the release version (e.g. `1.9.0-dev`).
> * To make a second pre-release version, manually update the version in `main`
  (e.g. update it from `1.9.0-dev` to `1.9.0rc2-dev`)
  before running the prepare release branch workflow for that release.

trask commented:
ok, I've re-tested all the workflows on my otel-python fork, I think this is ready for a final review

lzchen commented on RELEASING.md:
```diff
@@ -1,92 +1,68 @@
+
+## Preparing a new major or minor release
+
+* Run the [Prepare release branch workflow](https://github.com/trask/repository-template/actions/workflows/prepare-release-branch.yml).
```
Still pointing to your personal repo?

trask replied:
will fix

lzchen commented on RELEASING.md:
```diff
@@ -1,92 +1,68 @@
+* Run the [Release workflow](https://github.com/open-telemetry/opentelemetry-python/actions/workflows/release.yml).
+  * Press the "Run workflow" button, then select the release branch from the dropdown list,
+    e.g. `release/v1.9.x`, and click the "Run workflow" button below that.
+  * This workflow will publish the artifacts and publish a GitHub release with release notes based on the change log.
```
@trask 
For some reason, the `publish` workflow, which is responsible for publishing to PyPi, and which usually runs automatically when a release is published was not run on my fork when testing. Were you able to get this to run automatically on your fork?

trask replied:
oh good catch, I think the workflow isn't running b/c the release is being published via workflow using the default GITHUB_TOKEN, and github has some protections against workflows triggering other workflows to prevent infinite loops (which is one of the reasons why many of the workflows in the PR use the BOT_TOKEN instead of the default GITHUB_TOKEN.

is it ok if I embed the `publish` workflow inside the release workflow (after the release is published)? that would be the simplest fix

trask replied:
I embedded the publishing into the release workflow. There are ways to componentize large workflows, but that also brings some complexity.

trask commented:
> For the two PRs that are generated from `prepare-release-branch` workflow, for the second pr that updates the version in main from `1.12.0rc2-dev` to `1.12.0-dev`, is this expected? I'm not too familiar with `dev` version conventions but should the version be the NEXT version? i.e. `1.12.1-dev` or `1.13-dev`? Something like that?

In Java land, we use `-SNAPSHOT`, which has special meaning in maven (the primary Java build tool), and whatever version we are working towards, e.g. `1.12.0`, we append `-SNAPSHOT` to the version and our local build will get published as `1.12.0-SNAPSHOT`.

If we are working towards `1.12.0-rc.2`, then our local build would be `1.12.0-rc.2-SNAPSHOT`, and when we are ready to release then we remove the `-SNAPSHOT`, and release `1.12.0-rc.2`.

After releasing `1.12.0-rc.2`, in Java land, we often bump to `1.12.0-SNAPSHOT`, under the assumption that the next release will be the final `1.12.0` release (which comes after `1.12.0-rc.2` prerelease in semver ordering).

I guessed that `-dev` in python is similar to `-SNAPSHOT`, but I could definitely be off base.

Also I notice that your prerelease versions don't follow semver (e.g. `1.12.0rc1` instead of `1.12.0-rc1`), not sure if that's a python thing, or if you'd like to change that.

If you can describe the versioning strategy that you would like to follow, I can update the workflows accordingly.

And if you don't want to mess with `-dev` that's not a problem, I can remove that. I assume in that case you'd want the version in `main` to be the version you are working towards? (without and `-dev` suffix)

lzchen commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,43 @@
+    runs-on: ubuntu-latest
+    steps:
+      - run: |
+          if [[ ! $GITHUB_REF_NAME =~ ^release/v[0-9]+\.[0-9]+\.x-0\.[0-9]+bx$ ]]; then
```
For some reason, this regex isn't picking up my release branch named `release/v1.13.0-0.33b0`. Does the "long term branch" have to have "x" or is that a wildcard of some sort?

trask replied:
did you manually create the `release/v1.13.0-0.33b0` branch? the prepare-release-branch workflow should have named it `release/v1.13.x-0.33bx`, e.g. https://github.com/trask/opentelemetry-python/tree/release/v1.12.x-0.33bx

the naming is to indicate that the release branch covers all patch releases (e.g. v1.12.1-0.33b1 will be released from that same release branch)

lzchen replied:
You're right, it works as expected now.

trask commented:
> And if you don't want to mess with -dev that's not a problem, I can remove that. I assume in that case you'd want the version in main to be the ~version you are working towards~ prior version? (without and -dev suffix)

I looked at the opentelemetry-js and opentelemetry-go repos, and neither of them use `-dev`, so I'm thinking that's just a special Java thing (`-SNAPSHOT`).

In those two repos, they leave the version alone in the `main` branch until it's time to release and then they bump it. (so generally the version in `main` reflects the most recently released version).

Which looks like exactly what you are doing in this repo currently.

So I'm thinking it makes sense to remove the `-dev` stuff at least for now.

trask commented:
I kept the `-dev` suffix for now, but improved the prerelease ("release candidate") handling:

* renamed "prerelease" in the workflow to "unstable", I think this is a clearer distinction between "stable" components vs "unstable" components instead of "stable" vs "prerelease" since you can have prereleases (i.e. RCs) of stable components
* added an optional input parameter when running the "prepare release branch" workflow where you can specify a prerelease version (for the stable components), this improves a couple of things:
  * no longer need to manually update the version on `main` to `1.9.0rc1-dev` before running the workflow when you want to make a prerelease
  * no longer have the weird "prerelease of a prerelease" version (even in main) `1.9.0-rc1-dev`
  * so now the version in main remains `1.9.0-dev`, which is a (semver) prerelease of `1.9.0`, so makes more sense I think

trask commented:
I will do another full test of the workflows on my fork, but will wait a bit to see if you would like any other behavioral changes first

trask commented:
I switched the order to publish to PyPI before publishing the GitHub release.

This is what we've done in Java, since publishing to Maven is more likely to fail, but I can switch the ordering back if you'd like.

lzchen commented:
@trask 
> Also I notice that your prerelease versions don't follow semver (e.g. 1.12.0rc1 instead of 1.12.0-rc1), not sure if that's a python thing, or if you'd like to change that.

It is a python specific thing: https://peps.python.org/pep-0440/#pre-releases

lzchen commented:
@trask 

> So I'm thinking it makes sense to remove the -dev stuff at least for now.

I think adding `dev` is a decision made by the Python SIG because we won't have a way to conveniently test local changes. We would want to add that regardless of this workflow pr. However to not block this, you can assume that `dev` is not part of our workflow and we will fix the workflow once it is added.

trask commented:
> It is a python specific thing: https://peps.python.org/pep-0440/#pre-releases

nice, thx for sharing the link 👍

> I think adding `dev` is a decision made by the Python SIG because we won't have a way to conveniently test local changes

I see in the same PEP 440 they also define the versioning for [Developmental releases](https://peps.python.org/pep-0440/#developmental-releases), I will update the workflow to conform to that convention 👍

trask commented:
oh, I also created a new bot account @opentelemetrybot for everyone to use, and updated the workflow to use that

I will share the Personal Access Token with you, and I'll register the bot account at https://github.com/open-telemetry/community/blob/main/assets.md

srikanthccv commented on .github/workflows/prepare-release-branch.yml:
```diff
@@ -0,0 +1,177 @@
+        env:
+          PRERELEASE_VERSION: ${{ github.event.inputs.prerelease_version }}
+        run: |
+          if [[ $GITHUB_REF_NAME != main ]]; then
```
nit:

```suggestion
          if [[ $GITHUB_REF_NAME != ${{ github.event.repository.default_branch }}
 ]]; then
```

trask replied:
there are several places that refer to "main", I'm not sure it makes things clearer to change all references from "main" to "the default branch"

srikanthccv replied:
I was trying this on my fork with different default branch. So it was not successful at the start. I am fine with keeping it main.

trask commented:
I added documentation in `RELEASING.md` to better explain the version numbering:

> ## Notes about version numbering for stable components
> 
> * The version number for stable components in the `main` branch is always `X.Y.0.dev`,
>   where `X.Y.0` represents the next minor release.
> * When the release branch is created, you can opt to make a "pre-release", e.g. `X.Y.0rc2`.
> * If you ARE NOT making a "pre-release":
>   * A "long-term" release branch will be created, e.g. `release/v1.9.x-0.21bx` (notice the wildcard x's).
>     Later on, after the initial release, you can backport PRs to a "long-term" release branch and make patch releases
>     from it.
>   * The version number for stable components in the release branch will be bumped to remove the `.dev`,
>     e.g. `X.Y.0`.
>   * The version number for stable components in the `main` branch will be bumped to the next version,
>     e.g. `X.{Y+1}.0.dev`.
> * If you ARE making a "pre-release":
>   * A "short-term" release branch will be created, e.g. `release/v1.9.0rc2-0.21b0` (notice the precise version with no
>     wildcard x's). "Short-term" release branches do not support backports or patch releases after the initial release.
>  * The version number for stable components in the `main` branch will not be bumped, e.g. it will remain `X.Y.0.dev`
>     since the next minor release will still be `X.Y.0`.
> 
> ## Notes about version numbering for unstable components
> 
> * The version number for unstable components in the `main` branch is always `0.Yb0.dev`,
>   where `0.Yb0` represents the next minor release.
>   * _Question: Is "b" (beta) redundant on "0." releases, or is this a python thing? I'm wondering if we can change it to `0.Y.0` to match up with the practice in js and go repos._
> * Unstable components do not need "pre-releases", and so whether or not you are making a "pre-release" of stable
>   components:
>   * The version number for unstable components in the release branch will be bumped to remove the `.dev`,
>     e.g. `0.Yb0`.
>   * The version number for unstable components in the `main` branch will be bumped to the next version,
>     e.g. `0.{Y+1}b0.dev`.

srikanthccv commented:
@trask since we had released rc2 I think we can get this reviewed and merged, please resolve the conflicts.

trask commented:
merge conflicts resolved, ptal 👍

srikanthccv commented:
There have been many changes since the last time. Does it make sense to revive this now? I think we don't any other reason to put this on hold.

ocelotl commented:
> There have been many changes since the last time. Does it make sense to revive this now? I think we don't any other reason to put this on hold.

So, you see value in reviving this, @srikanthccv? (I do)

trask commented:
Rebased and re-tested the release process (both minor and patch) in my fork.

I realized that the backport automation (for patches) will fail currently most of the time due to change log conflicts. I think I can make the backport automation smarter and handle that, but I don't think it's necessarily required for landing this PR, as backports for patch releases can still be done manually.

srikanthccv reviewed (approved):
Thanks, We will need to do this for contrib as well, not necessarily immediately.

srikanthccv commented:
Ah I see the failed jobs are because the contrib versions are not `.dev` and fail to work. I think it's going to some effort to create same for contrib. We can mark not required for the time being but I would like to see that addressed soon as well. What do you think @lzchen

lzchen commented:
@srikanthccv 
When you say marked not required, do you mean marking the contrib tests as not required for merging? I am fine with that for now. As for how to fix the builds in the near future, it should simply be marking the contrib versions with `dev` right?

srikanthccv commented:
> When you say marked not required, do you mean marking the contrib tests as not required for merging? I am fine with that for now

I was talking about the `trace-context` job, the `contrib-build` tests are already not marked required. 

> As for how to fix the builds in the near future, it should simply be marking the contrib versions with dev right?

Yes.
